cli/coverage.rs

```diff
@@ -0,0 +1,250 @@
+    new_box_with(move |self_ptr| {
+      let v8_channel = v8::inspector::ChannelBase::new::<Self>();
+      let v8_session = unsafe { &mut *inspector_ptr }.connect(
+        Self::CONTEXT_GROUP_ID,
+        unsafe { &mut *self_ptr },
+        v8::inspector::StringView::empty(),
+      );
+
+      let (sender, receiver) = mpsc::channel::<String>(1);
+
+      Self {
+        v8_channel,
+        v8_session,
+        sender,
+        receiver,
+      }
+    })
+  }
+
+  async fn dispatch(&mut self, message: String) -> Result<String, ErrBox> {
+    let message = v8::inspector::StringView::from(message.as_bytes());
+    self.v8_session.dispatch_protocol_message(message);
+
+    let response = self.receiver.try_next()?;
+    Ok(response.unwrap())
+  }
+
+  pub async fn start_collecting(&mut self) -> Result<(), ErrBox> {
+    self
+      .dispatch(r#"{"id":1,"method":"Runtime.enable"}"#.into())
+      .await?;
+    self
+      .dispatch(r#"{"id":2,"method":"Profiler.enable"}"#.into())
+      .await?;
+
+    self
+        .dispatch(r#"{"id":3,"method":"Profiler.startPreciseCoverage", "params": {"callCount": true, "detailed": true}}"#.into())
+        .await?;
+
+    Ok(())
+  }
+
+  pub async fn take_precise_coverage(
+    &mut self,
+  ) -> Result<Vec<ScriptCoverage>, ErrBox> {
+    let response = self
+      .dispatch(r#"{"id":4,"method":"Profiler.takePreciseCoverage" }"#.into())
+      .await?;
+
+    let coverage_result: TakePreciseCoverageResponse =
+      serde_json::from_str(&response).unwrap();
+
+    Ok(coverage_result.result.result)
+  }
+
+  pub async fn stop_collecting(&mut self) -> Result<(), ErrBox> {
+    self
+      .dispatch(r#"{"id":5,"method":"Profiler.stopPreciseCoverage"}"#.into())
+      .await?;
+
+    self
+      .dispatch(r#"{"id":6,"method":"Profiler.disable"}"#.into())
+      .await?;
+
+    self
+      .dispatch(r#"{"id":7,"method":"Runtime.disable"}"#.into())
+      .await?;
+
+    Ok(())
+  }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CoverageRange {
+  pub start_offset: usize,
+  pub end_offset: usize,
+  pub count: usize,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct FunctionCoverage {
+  pub function_name: String,
+  pub ranges: Vec<CoverageRange>,
+  pub is_block_coverage: bool,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ScriptCoverage {
+  pub script_id: String,
+  pub url: String,
+  pub functions: Vec<FunctionCoverage>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct TakePreciseCoverageResult {
+  result: Vec<ScriptCoverage>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct TakePreciseCoverageResponse {
+  id: usize,
+  result: TakePreciseCoverageResult,
+}
+
+pub struct PrettyCoverageReporter {}
+
+impl PrettyCoverageReporter {
+  pub fn new() -> PrettyCoverageReporter {
+    PrettyCoverageReporter {}
+  }
+
+  pub fn visit(
```

Comment:
Could you split this structure into two separate structures? One for calculating coverage, etc. The other for actual printing to terminal.